DevCycle/Models/EventQueue.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  EventQueue.swift
+//  DevCycle
+//
+//  Copyright © 2022 Taplytics. All rights reserved.
+//
+
+import Foundation
+
+enum EventQueueErrors: Error {
+    case FlushingInProgress
+}
+
+class EventQueue {
+    var eventDispatchQueue = DispatchQueue(label: "com.devcycle.EventQueue")
+    var events: [DVCEvent] = []
+    var aggregateEventQueue: DVCAggregateEvents = DVCAggregateEvents()
+    var flushing: Bool = false
+    
+    func queue(_ event: DVCEvent) {
+        eventDispatchQueue.async {
+            self.events.append(event)
+        }
+    }
+    
+    func queue(_ events: [DVCEvent]) {
+        eventDispatchQueue.async {
+            self.events.append(contentsOf: events)
+        }
+    }
+    
+    func flush(service: DevCycleServiceProtocol, user: DVCUser, callback: FlushCompletedHandler? = nil) {
+        if (self.flushing) {
+            Log.warn("Flushing already in progress, cancelling flush")
+            callback?(EventQueueErrors.FlushingInProgress)
+            return
+        }
+        
+        var eventsToFlush: [DVCEvent] = []
+        eventDispatchQueue.sync {
+            self.flushing = true
+            eventsToFlush = self.events
+            eventsToFlush.append(contentsOf: self.aggregateEventQueue.variableDefaulted.map { (_: String, defaultedEvent: DVCEvent) -> DVCEvent in
```

Comment:
Is there a way to iterate over the keys in the `self.aggregateEventQueue` map? as ideally we don't want to have to edit this when other event types are added.

Reply:
we can do that